src/pymedphys/gamma/_level2/gammashell.py

```diff
@@ -116,160 +114,290 @@ def gamma_shell(coords_reference, dose_reference,
+    for i, dose_threshold in enumerate(options.dose_percent_threshold):
+        for j, distance_threshold in enumerate(options.distance_mm_threshold):
+            key = (dose_threshold, distance_threshold)
+
+            gamma_temp = current_gamma[:, i, j]
+            gamma_temp = np.reshape(
+                gamma_temp, np.shape(dose_reference))
+            gamma_temp[np.isinf(gamma_temp)] = np.nan
+
+            with np.errstate(invalid='ignore'):
+                gamma_greater_than_ref = gamma_temp > max_gamma
+                gamma_temp[gamma_greater_than_ref] = max_gamma
+
+            gamma[key] = gamma_temp
+
+    if not options.quiet:
+        print('\nComplete!')
+
+    if len(gamma.keys()) == 1:
+        gamma = next(iter(gamma.values()))
+
+    return gamma
+
+
+def default_ram() -> int:
+    memory = psutil.virtual_memory()
+    total_memory = memory.total
+
+    ram_available = int(total_memory * 0.8)
+
+    return ram_available
+
+
+def expand_dims_to_1d(array):
+    array = np.array(array)
+    dims = len(np.shape(array))
+
+    if dims == 0:
+        return array[None]
 
-    dose_reference = np.array(dose_reference)
-    reference_dose_above_threshold = dose_reference >= lower_dose_cutoff
+    if dims == 1:
+        return array
 
-    mesh_coords_reference = np.meshgrid(*coords_reference, indexing='ij')
-    flat_mesh_coords_reference = [
-        np.ravel(item)
-        for item in mesh_coords_reference]
+    raise ValueError("Expected a 0-d or 1-d array")
 
-    if mask_evaluation:
-        coordinates_at_distance_shell = calculate_coordinates_shell(
-            0, len(coords_reference), distance_step_size)
 
-        interpolated_evaluation_dose = interpolate_evaluation_dose_at_distance(
-            evaluation_interpolation, flat_mesh_coords_reference,
-            coordinates_at_distance_shell)[0, :]
+@dataclass(frozen=True)
+class GammaInternalFixedOptions():
+    flat_mesh_coords_reference: np.ndarray
+    flat_dose_reference: np.ndarray
+    reference_points_to_calc: np.ndarray
+    dose_percent_threshold: np.ndarray
+    distance_mm_threshold: np.ndarray
+    evaluation_interpolation: RegularGridInterpolator
+    interp_fraction: int
+    max_gamma: float
+    lower_dose_cutoff: float = 0
+    maximum_test_distance: float = np.inf
+    global_normalisation: Optional[float] = None
+    local_gamma: bool = False
+    skip_once_passed: bool = False
+    ram_available: Optional[int] = None
+    quiet: bool = False
 
-        interpolated_evaluation_dose = np.reshape(
-            interpolated_evaluation_dose, np.shape(dose_reference))
+    def __post_init__(self):
+        self.set_defaults()
 
-        evaluation_dose_interpolation_within_bounds = (
-            interpolated_evaluation_dose != np.inf)
-        evaluation_dose_above_threshold = (
-            interpolated_evaluation_dose >= lower_dose_cutoff)
+    def set_defaults(self):
+        if self.global_normalisation is None:
+            object.__setattr__(
```

Comment:
It's actually because I have explicitly frozen this object. The idea is once that object is instantiated none of it's variables should change. Keeps the logic of the object far simpler.

Reply:
👍